torch/utils/data/dataloader.py

```diff
@@ -95,11 +100,13 @@ def _worker_loop(dataset, index_queue, data_queue, collate_fn, seed, init_fn, wo
         try:
             r = index_queue.get(timeout=MANAGER_STATUS_CHECK_INTERVAL)
         except queue.Empty:
-            if watchdog.is_alive():
+            if watchdog.is_alive() and not done_event.is_set():
                 continue
             else:
                 break
-        if r is None:
+        # use done_event so that we can get faster exiting signal even if there
+        # are still indices in index_queue
+        if r is None or done_event.is_set():
```

Comment:
Does this serve any purpose other than to make them exit faster? If yes, I'd rather not use it, as it complicates the whole protocol for no good reason.

Reply:
It isn't necessary to fix the hang. But I'd argue that we should have this. Without this, the worker exiting is delayed until all indices in the queue are processed. This could 
1. be unexpected by users thus problematic if `dataset.__getitem__` uses scarce resources (e.g. gpu), and
3. cause `__del__` to be slow if the `dataset.__getitem__` is expensive. 

Moreover, `done_event` is already used in addition to `None` in exiting the `manager_thread`, so I think it is fine.

Reply:
Fair enough. If we've already been using this for the pinning thread, then we can improve the speed in here as well.